### PR TITLE
Time - Kate P.

### PR DESCRIPTION
**Pull request by Catherina87**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | One of the custom methods I wrote is the top_movies. This method calls sort_movie_works method which sorts the works by movie category and sorts them in descending order. Then top_movies method takes the first ten movies (these are the top tens). 
Describe how you approached testing that model method. What edge cases did you come up with? | I'm testing that the method should return 10 top movies; then returns 5 movies if there are only 5 of them in total; the edge case is that the method should return empty array if there are no movies.
What are `session` and `flash`? What is the difference between them? | "session" and "flash" are both hashed that store some date. The "flash" only keeps data from request to request, while the "session" keeps data throughout the time when user is logged in. Session data can be deleted if we clean cookies.
What was one thing that you gained more clarity on through this assignment? | Deploying to Heroku, code flow, migrations.
What is the Heroku URL of your deployed application? | https://media-ranker-2020-katty.herokuapp.com/

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+Rails.application.routes.draw do
+
+  get "top_works/index"
```
Comment (CheezItMan) [inline config/routes.rb:3]: You don't need this route if it's being used as the root path.

File: config/routes.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  root to: 'top_works#index'
+  resources :works
```
Comment (CheezItMan) [inline config/routes.rb:5]: You've defined the `resources :works` twice, also on line 13.

```suggestion
```

File: test/models/work_test.rb [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+        works = Work.top_movies
+        expect(works[0].title).must_equal works(:movie_two).title
+      end
+
+      it "returns empty array if there are no movies" do 
+        movies = Work.sort_movie_works
+        movies.each do |m|
+          m.category = "book"
+          m.save
+        end
```
Comment (CheezItMan) [inline test/models/work_test.rb:323]: Instead of reassigning all the categories here you could simply do `Work.destroy_all`  Or you could do.

`Work.where(category: 'movie').destroy_all` to just destroy all the movies.

File: test/models/vote_test.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        work_id: works(:titanic).id,
+        user_id: users(:user_one).id,
+        date: Date.today
+      )
+
+      expect(vote.user.username).must_equal users(:user_one).username
+    end
+  end
+
+  describe "validations" do
```
Comment (CheezItMan) [inline test/models/vote_test.rb:46]: You should also have a validation that ensures that the combination of user_id and work_id are unique.

File: test/controllers/works_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe WorksController do
+  # it "does a thing" do
```
Comment (CheezItMan) [inline test/controllers/works_controller_test.rb:4]: Just noting there  are no controller tests.

File: app/views/works/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+          <th>Created By</th>
+          <th>Published</th>
+          <th>Upvote</th>
+        </tr>
+      </thead>
+      <tbody>
+
+        <% movies = Work.sort_movie_works%>
+
+        <% movies.each do |movie| %>
```
Comment (CheezItMan) [inline app/views/works/index.html.erb:95]: Just noting that these things could be done with a set of partial views.  You could build this view and use the same partial view 3 times.

File: app/models/work.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  def self.sort_book_works
+    return Work.where(category: "book").left_joins(:votes).group(:id).order('COUNT(votes.id) DESC')
+  end
+
+  def self.sort_movie_works
+    return Work.where(category: "movie").left_joins(:votes).group(:id).order('COUNT(votes.id) DESC')
+  end
+
```
Comment (CheezItMan) [inline app/models/work.rb:25]: Notice that these methods are pretty much the same.  You could instead use one method with a parameter.

File: app/models/vote.rb [hunk 8]
```diff
@@ -0,0 +1,7 @@
+class Vote < ApplicationRecord
+  belongs_to :work
+  belongs_to :user
+
+  validates :work_id, presence: true
```
Comment (CheezItMan) [inline app/models/vote.rb:5]: The relationship will already make the `work_id` a required field. 

```suggestion
```

File: app/controllers/works_controller.rb [hunk 9]
```diff
@@ -0,0 +1,74 @@
+class WorksController < ApplicationController
+  before_action :find_work, only: [:show, :edit, :update, :destroy]
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:2]: 👍

**Comment by CheezItMan** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ⚠ An extra route and duplicate resources.
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️, it has the method
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅


## Summary


Well done.  You hit the learning goals here.  I like how you styled the site to look like the original.  This demonstrates a solid understanding of basic Rails.  Take a look at my comments and let me know where you have questions.
